File: tests/testdata/codegen/components-pp/go/exampleComponent.go
package main

import (
	"fmt"

	"github.com/pulumi/pulumi-random/sdk/v4/go/random"
	"github.com/pulumi/pulumi/sdk/v3/go/pulumi"
	"github.com/pulumi/pulumi/sdk/v3/go/pulumi/config"
)

type DeploymentZonesArgs struct {
	Zone pulumi.StringInput
}

type GithubAppArgs struct {
	Id            pulumi.StringInput
	KeyBase64     pulumi.StringInput
	WebhookSecret pulumi.StringInput
}

type ServersArgs struct {
	Name pulumi.StringInput
}

type ExampleComponentArgs struct {
	Input           pulumi.StringInput
	CidrBlocks      map[string]pulumi.StringInput
	GithubApp       *GithubAppArgs
	Servers         []*ServersArgs
	DeploymentZones map[string]*DeploymentZonesArgs
	IpAddress       []pulumi.IntInput
}

type ExampleComponent struct {
	pulumi.ResourceState
	Result pulumi.AnyOutput
}

func NewExampleComponent(
	ctx *pulumi.Context,
	name string,
	args *ExampleComponentArgs,
	opts ...pulumi.ResourceOption,
) (*ExampleComponent, error) {
	var componentResource ExampleComponent
	err := ctx.RegisterComponentResource("components:index:ExampleComponent", name, &componentResource, opts...)
	if err != nil {
		return nil, err
	}
	password, err := random.NewRandomPassword(ctx, fmt.Sprintf("%s-password", name), &random.RandomPasswordArgs{
		Length:          pulumi.Int(16),
		Special:         true,
		OverrideSpecial: args.Input,
	}, pulumi.Parent(&componentResource))
	if err != nil {
		return nil, err
	}
	_, err = random.NewRandomPassword(ctx, fmt.Sprintf("%s-githubPassword", name), &random.RandomPasswordArgs{
		Length:          pulumi.Int(16),
		Special:         true,
		OverrideSpecial: args.GithubApp.WebhookSecret,
	}, pulumi.Parent(&componentResource))
	if err != nil {
		return nil, err
	}
	// Example of iterating a list of objects
	var serverPasswords []*random.RandomPassword
	for index := 0; index < len(args.Servers); index++ {
		key0 := index
		val0 := index
		__res, err := random.NewRandomPassword(ctx, fmt.Sprintf("%s-serverPasswords-%v", name, key0), &random.RandomPasswordArgs{
			Length:          pulumi.Int(16),
			Special:         true,
			OverrideSpecial: pulumi.String(args.Servers[val0].Name),
		}, pulumi.Parent(&componentResource))
		if err != nil {
			return nil, err
		}
		serverPasswords = append(serverPasswords, __res)
	}
	// Example of iterating a map of objects
	var zonePasswords []*random.RandomPassword
	for key0, val0 := range args.DeploymentZones {
		__res, err := random.NewRandomPassword(ctx, fmt.Sprintf("%s-zonePasswords-%v", name, key0), &random.RandomPasswordArgs{
			Length:          pulumi.Int(16),
			Special:         true,
			OverrideSpecial: pulumi.String(val0),
		}, pulumi.Parent(&componentResource))
		if err != nil {
			return nil, err
		}
		zonePasswords = append(zonePasswords, __res)
	}
	_, err = NewSimpleComponent(ctx, fmt.Sprintf("%s-simpleComponent", name), nil, pulumi.Parent(&componentResource))
	if err != nil {
		return nil, err
	}
	err = ctx.RegisterResourceOutputs(&componentResource, pulumi.Map{
		"result": password.Result,
	})
	if err != nil {
		return nil, err
	}
	componentResource.Result = password.Result
	return &componentResource, nil
}

File: tests/testdata/codegen/components-pp/python/exampleComponent.py
import pulumi
from pulumi import Input
from simpleComponent import SimpleComponent
from typing import Optional, Dict, TypedDict, Any
from typing_extensions import NotRequired
import builtins as _builtins
import pulumi_random as random

class DeploymentZones(TypedDict):
    zone: NotRequired[Input[_builtins.str]]

class GithubApp(TypedDict):
    id: NotRequired[Input[_builtins.str]]
    keyBase64: NotRequired[Input[_builtins.str]]
    webhookSecret: NotRequired[Input[_builtins.str]]

class Servers(TypedDict):
    name: NotRequired[Input[_builtins.str]]

class ExampleComponentArgs(TypedDict):
    input: Input[_builtins.str]
    cidrBlocks: Input[Dict[_builtins.str, _builtins.str]]
    githubApp: NotRequired[Input[GithubApp]]
    servers: NotRequired[Input[list(Servers)]]
    deploymentZones: NotRequired[Input[Dict[str, DeploymentZones]]]
    ipAddress: Input[list[_builtins.int]]

class ExampleComponent(pulumi.ComponentResource):
    def __init__(self, name: str, args: ExampleComponentArgs, opts:Optional[pulumi.ResourceOptions] = None):
        super().__init__("components:index:ExampleComponent", name, args, opts)

        password = random.RandomPassword(f"{name}-password",
            length=16,
            special=True,
            override_special=args["input"],
            opts = pulumi.ResourceOptions(parent=self))

        github_password = random.RandomPassword(f"{name}-githubPassword",
            length=16,
            special=True,
            override_special=args["githubApp"]["webhookSecret"],
            opts = pulumi.ResourceOptions(parent=self))

        # Example of iterating a list of objects
        server_passwords: list[random.RandomPassword] = []
        for server_passwords_range in [{"value": i} for i in range(0, len(args["servers"]))]:
            server_passwords.append(random.RandomPassword(f"{name}-serverPasswords-{server_passwords_range['value']}",
                length=16,
                special=True,
                override_special=args["servers"][server_passwords_range["value"]]["name"],
                opts = pulumi.ResourceOptions(parent=self)))

        # Example of iterating a map of objects
        zone_passwords: dict[str, random.RandomPassword] = {}
        for zone_passwords_range in [{"key": k, "value": v} for [k, v] in sorted((args["deploymentZones"]).items())]:
            zone_passwords[zone_passwords_range['key']] = random.RandomPassword(f"{name}-zonePasswords-{zone_passwords_range['key']}",
                length=16,
                special=True,
                override_special=zone_passwords_range["value"]["zone"],
                opts = pulumi.ResourceOptions(parent=self))

        simple_component = SimpleComponent(f"{name}-simpleComponent", opts = pulumi.ResourceOptions(parent=self))

        self.result = password.result
        self.register_outputs({
            'result': password.result
        })